### PR TITLE
Allow bindings commands to context menu actions.

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This allows binding Lisp commands to context menu actions with the help of the new `ffi-add-context-menu-command`. Those actions are only available when the respecitve command is (which means always for global commands, and only when the mode is enabled—for mode-specific ones). It's quite basic, which might be a good reason it works in the first place xD

# Discussion

- [ ] Any idea for a better design? Maybe include this functionality in `define-command` somehow?
- [ ] Now that I think of it, internal schemes and `define-internal-scheme` should also have an `ffi-` prefix, shouldn't they? Or maybe we should provide the most intuitive interface possible, while doing the renderer-specific things based on internal information? In this case propagating context menu actions inside `define-*command-*` makes lots of sense...
- [X] I've updated `cl-webkit` to 3.5.4, incorporating all the recent useful changes. @Ambrevar, care to send a patch to Guix?

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [X] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [x] and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - [X] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [X] I have updated the existing documentation to match my changes.
  - [X] I have commented my code in hard-to-understand areas.
  - [X] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [X] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
  - [X] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [X] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by aartaka** [comment]
Just to mention, WebKit context menu items:
- Can be checkboxes or radios.
- Can be grouped.
- Can access DOM/widget they are pointing at.
- Get enabled/disabled given certain conditions (like pointing at a textarea or a video).
- Can store arbitrary data and modify it besed on item presses.

None of this is immediately available to items introduced in this PR, but then, those are full-blown Nyxt commands, so maybe we don't need any of this shiny styff...

**Comment by aadcg** [comment]
@Ambrevar, @aartaka I've sent the patch to Guix (it's bug #57603).

**Comment by Ambrevar** [comment]
It's already merged :)

**Comment by aartaka** [comment]
So, anyone willing to review this? Shall I merge?

File: source/foreign-interface.lisp [hunk 1]
```diff
@@ -363,3 +363,20 @@ Return the text cut."))
...
   (:documentation "Redo the last undone text edit performed in BUFFER's web view."))
+
+(defvar *context-menu-commands* (make-hash-table)
+  "A hash table from command symbols to context menu labels for those.
+Once a context menu appears, those commands will be added to it as actions with
+the labels they have as hash values.")
+
+(define-ffi-generic ffi-add-context-menu-command (command &optional label)
+  (:method :around ((command command) &optional label)
+    (call-next-method command (or label (format nil "~:(~a~)" (str:remove-punctuation (symbol-name (name command)))))))
```
Comment (aadcg) [inline source/foreign-interface.lisp:374]: Is `format` worth it here? It's a bit hard to follow.

Perhaps use `str:sentence-case` and `str:remove-punctuation`?

File: source/renderer/gtk.lisp [hunk 2]
```diff
@@ -1583,7 +1583,31 @@ See `finalize-buffer'."
                   (webkit:webkit-context-menu-item-new-from-stock-action-with-label
                    :webkit-context-menu-action-open-link-in-new-window
                    "Open Link in New Buffer")
-                  i)))))))
+                  i))))))
+      (webkit:webkit-context-menu-append
+       context-menu (webkit:webkit-context-menu-item-new-separator))
```
Comment (aadcg) [inline source/renderer/gtk.lisp:1588]:
```
(webkit:webkit-context-menu-append context-menu
                                   (webkit:webkit-context-menu-item-new-separator))
```

Perhaps the indentation above would be easier to follow.

File: source/renderer/gtk.lisp [hunk 3]
```diff
@@ -1583,7 +1583,31 @@ See `finalize-buffer'."
...
+      (sera:and-let* ((_ (plusp (hash-table-count *context-menu-commands*)))
```
Comment (aadcg) [inline source/renderer/gtk.lisp:1589]: Perhaps `alex:positive-integer-p`? `plusp` checks for reals.

Comment (aartaka) [reply source/renderer/gtk.lisp:1589]: `plusp` does check for reals, yes, but `hash-table-count` returns solid non-negative integers, so `plusp` is likely to be optimized into a non-negative-integer-only version here.

File: source/renderer/gtk.lisp (hunk 2 above)
Comment (aartaka) [reply source/renderer/gtk.lisp:1588]: I usually optimize for minimal absolute indentation, so my version appeals to myself much more tha yours which makes the next indentation level to be extremely far from the current one.

But there's no answer here, I guess ¯\\\_(ツ)_/¯

File: source/renderer/gtk.lisp [hunk 4]
```diff
@@ -1583,7 +1583,31 @@ See `finalize-buffer'."
...
+                      (accessible-commands
+                       (mapcar #'name
+                               (list-commands
+                                :global-p t
+                                :mode-symbols (mapcar #'sera:class-name-of
```
Comment (aadcg) [inline source/renderer/gtk.lisp:1594]: The pattern `(mapcar #'sera:class-name-of foo` seems to appear 8 times in the codebase. Perhaps it should be abstracted?

File: source/renderer/gtk.lisp [hunk 5]
```diff
@@ -1583,7 +1583,31 @@ See `finalize-buffer'."
...
+                                                      (sera:filter #'enabled-p (modes buffer))))))
+                      (context-commands (alex:hash-table-keys *context-menu-commands*))
+                      (accessible-context-commands (intersection accessible-commands context-commands)))
```
Comment (aadcg) [inline source/renderer/gtk.lisp:1597]: I'd leave this binding out and would jump straight into:

`(dolist (command  (intersection accessible-commands context-commands)) ...`

File: source/renderer/gtk.lisp [hunk 6]
```diff
@@ -1583,7 +1583,31 @@ See `finalize-buffer'."
...
+        (dolist (command accessible-context-commands)
+          (let* ((label (gethash command *context-menu-commands*))
```
Comment (aadcg) [inline source/renderer/gtk.lisp:1599]: I'd leave the `label` binding out as well.

File: source/renderer/gtk.lisp [hunk 7]
```diff
@@ -1583,7 +1583,31 @@ See `finalize-buffer'."
...
+                 ;; Using stock actions here, because cl-cffi-gtk has a
+                 ;; terrible API for GActions, requiring an exact type to be
+                 ;; passed and disallowing NULL as a type :/
+                 (item (webkit:webkit-context-menu-item-new-from-stock-action-with-label
+                        :webkit-context-menu-action-action-custom label)))
+            (gobject:g-signal-connect
+             (webkit:webkit-context-menu-item-get-g-action item) "activate"
+             (lambda (action parameter)
+               (declare (ignore action parameter))
+               (run-async command)))
```
Comment (aadcg) [inline source/renderer/gtk.lisp:1609]:
```lisp
(gobject:g-signal-connect (webkit:webkit-context-menu-item-get-g-action item)
                                      "activate"
                                      (lambda (action parameter)
                                        (declare (ignore action parameter))
                                        (run-async command)))
```

I find the above indentation easier to follow.

**Comment by aadcg** [comment]
Very useful functionality. Regarding design:

> Maybe include this functionality in define-command somehow?

That seems clear, in theory.  If you have an idea right away, you could try it.  Otherwise, I think it's absolutely fine to open an issue and leave it for later!

File: source/renderer/gtk.lisp (hunk 4 above)
Comment (aartaka) [reply source/renderer/gtk.lisp:1594]: Maybe, but at this point we can only abstract towards APL...

**Comment by aartaka** [comment]
Let's leave it for later (⌒_⌒;)

**Comment by aartaka** [comment]
The indentation I'm not changing ヽ(ｏ`皿′ｏ)ﾉ

**Comment by aadcg** [comment]
> Let's leave it for later (⌒_⌒;)

Totally agree. There's too much going on... :woozy_face:

**Comment by aadcg** [comment]
> The indentation I'm not changing ヽ(ｏ`皿′ｏ)ﾉ

I'm only pushing for conventions, whatever the choice might be :)
